fix(webgl): normalize nearly identical vertices before tessellation

avinxshKD opened the pull request:
Resolves #8186

## Changes

This PR fixes tessellation artifacts that occur when drawing shapes with multiple contours where consecutive vertices have nearly identical coordinates (differing by ~1e-8 or similar small amounts). This commonly happens when using `font.textToContours()`.


- **Added visual test in `test/unit/visual/cases/webgl.js`:**
  - New visual test "Tessellation → Handles nearly identical consecutive vertices"
  - Easier for contributors to debug visually when issues arise
### Implementation Details

- **Modified `_tesselateShape()` in `src/webgl/p5.RendererGL.Immediate.js`:**
  - Added coordinate normalization loop before passing vertices to libtess tessellator
  - Consecutive vertices with coordinates within `1e-6` (epsilon) of each other are normalized to use the exact same value
  - Prevents numerical precision issues in libtess that cause glitchy rendering

- **Added comprehensive unit test in `test/unit/webgl/p5.RendererGL.js`:**
  - Tests tessellation with nearly identical consecutive vertices
  - Verifies that contour holes are properly cut out
  - Confirms shape fills work correctly with problematic coordinates

### Technical Background

This workaround addresses a numerical precision issue in libtess, which is a JavaScript port of SGI C code from the 1990s. When consecutive vertices have coordinates that are almost (but not exactly) equal, libtess produces glitchy tessellation output. This issue is particularly likely to occur with contours automatically sampled from fonts via `font.textToContours()`.

### Screenshots of the change

**Before:**
- Tessellation produces visual artifacts and incorrect geometry
- Referenced in issue #8186

**After (with this fix):**
- Clean tessellation with proper contour cutouts
- No visual artifacts

## PR Checklist

- [x] `npm run lint` passes
- [x] [Inline reference] is included / updated (N/A - internal fix, no API changes)
- [x] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

avinxshKD committed:
fix(webgl): normalize nearly identical vertices before tessellation

davepagurek reviewed (commented):
Hi, thanks for taking this on! ~~The approach looks good. I was initially testing in p5 2.x, have we confirmed that this issue exists in 1.x?~~ (Update: this exists in 1.x too, so we can keep this PR into `main` around, but we'll need another one into the `dev-2.0` branch for 2.x.) Regardless, it may be a good idea to start by making the change just in 2.x, in the `dev-2.0` branch instead of `main`.

avinxshKD committed:
refactor(test): convert tessellation test to visual test

Converted pixel-checking unit test to visual test as suggested by maintainer.
This makes the test easier to debug when issues arise in the future.

avinxshKD commented:
> Hi, thanks for taking this on! ~The approach looks good. I was initially testing in p5 2.x, have we confirmed that this issue exists in 1.x?~ (Update: this exists in 1.x too, so we can keep this PR into `main` around, but we'll need another one into the `dev-2.0` branch for 2.x.) Regardless, it may be a good idea to start by making the change just in 2.x, in the `dev-2.0` branch instead of `main`.

Thanks for the review @davepagurek

I've converted the pixel-checking unit test to a visual test as you suggested.

And also created a separate PR targeting the dev-2.0 branch with the same fixes, since the file structure is different there (ShapeBuilder.js instead of p5.RendererGL.Immediate.js). Here's the link https://github.com/processing/p5.js/pull/8221

Thanks

avinxshKD committed:
fix(webgl): remove z-axis normalization and update tessellation test

Removed Z-axis coordinate normalization as testing confirmed it's not needed.
Replaced simple test case with actual textToContours() example from issue #8186
that clearly demonstrates the tessellation bug and verifies the fix works.

Resolves #8186

avinxshKD committed:
fix(webgl): rewrite visual test, clean up vertex normalization

avinxshKD commented:
Hey @perminder-17 @davepagurek  @ksen0 updated the visual test, now uses hardcoded vertex data with a ~1e-8 coordinate offset that actually triggers the libtess bug. Old test was loading a font that didn't exist in test assets anyway.

Also cleaned up the normalization loop and removed the extra screenshots that got committed by mistake

<img width="578" height="618" alt="image" src="https://github.com/user-attachments/assets/bba53bdf-59be-494d-a7b7-9a885ec325cc" />

.

<img width="573" height="690" alt="image" src="https://github.com/user-attachments/assets/96c27b45-fd9b-4767-94a3-32f13a3ba442" />

avinxshKD commented:
Hey @davepagurek @perminder-17 can you pls look into this, thanks.

perminder-17 commented on test/unit/visual/cases/webgl.js:
```diff
@@ -141,4 +141,35 @@ visualSuite('WebGL', function() {
     });
   });
+
+  visualSuite('Tessellation', function() {
```
I think this comment is still unresolved:

it should be same as this; https://github.com/processing/p5.js/pull/8221/changes 

But I think in v1 there would be no addition of new features, so it would be safe to merge. Thanks for your input.

perminder-17 reviewed (approved):
Looks good to me!
